File: app/src/main/java/knf/kuma/emision/EmisionFragment.java
package knf.kuma.emision;

import android.os.AsyncTask;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import butterknife.BindView;
import butterknife.ButterKnife;
import knf.kuma.R;
import knf.kuma.custom.GridRecyclerView;
import knf.kuma.database.CacheDB;
import knf.kuma.database.dao.AnimeDAO;
import knf.kuma.pojos.AnimeObject;
import knf.kuma.widgets.emision.WEmisionProvider;
import pl.droidsonroids.jspoon.Jspoon;

public class EmisionFragment extends Fragment {
    @BindView(R.id.recycler)
    GridRecyclerView recyclerView;
    @BindView(R.id.error)
    View error;
    @BindView(R.id.progress)
    ProgressBar progressBar;
    private AnimeDAO dao = CacheDB.INSTANCE.animeDAO();
    private EmisionAdapter adapter;
    private boolean isFirst = true;

    public EmisionFragment() {
    }

    public static EmisionFragment get(AnimeObject.Day day) {
        Bundle bundle = new Bundle();
        bundle.putInt("day", day.value);
        EmisionFragment fragment = new EmisionFragment();
        fragment.setArguments(bundle);
        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = LayoutInflater.from(getContext()).inflate(R.layout.recycler_emision, container, false);
        ButterKnife.bind(this, view);
        adapter = new EmisionAdapter(this);
        recyclerView.setAdapter(adapter);
        return view;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        if (getContext() != null)
            CacheDB.INSTANCE.animeDAO().getByDay(getArguments().getInt("day", 1), getBlacklist()).observe(this, animeObjects -> {
                progressBar.setVisibility(View.GONE);
                if (isFirst && animeObjects != null && animeObjects.size() != 0) {
                    isFirst = false;
                    adapter.update(animeObjects);
                    recyclerView.scheduleLayoutAnimation();
                    checkStates(animeObjects);
                }
                error.setVisibility((animeObjects == null || animeObjects.size() == 0) ? View.VISIBLE : View.GONE);
            });
    }

    private void checkStates(final List<AnimeObject> animeObjects) {
        AsyncTask.execute(() -> {
            try {
                for (AnimeObject animeObject : animeObjects) {
                    try {
                        Document document = Jsoup.connect(animeObject.link).cookie("device", "computer").get();
                        AnimeObject object = new AnimeObject(animeObject.link, Jspoon.create().adapter(AnimeObject.WebInfo.class).fromHtml(document.outerHtml()));
                        if (!object.state.equals("En emisión")) {
                            dao.updateAnime(object);
                            adapter.remove(adapter.list.indexOf(animeObject));
                            WEmisionProvider.update(getContext());
                        }
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
            } catch (Exception e) {
                e.printStackTrace();
            }
        });
    }

    void reloadList() {
        if (getContext() != null)
            CacheDB.INSTANCE.animeDAO().getByDay(getArguments().getInt("day", 1), getBlacklist()).observe(this, animeObjects -> {
                error.setVisibility(View.GONE);
                if (animeObjects != null && animeObjects.size() != 0)
                    adapter.update(animeObjects);
                else adapter.update(new ArrayList<>());
                if (animeObjects == null || animeObjects.size() == 0)
                    error.setVisibility(View.VISIBLE);
            });
    }

    private Set<String> getBlacklist() {
        if (PreferenceManager.getDefaultSharedPreferences(getContext()).getBoolean("show_hidden", false))
            return new LinkedHashSet<>();
        else
            return PreferenceManager.getDefaultSharedPreferences(getContext()).getStringSet("emision_blacklist", new LinkedHashSet<>());
    }
}

File: app/src/main/java/knf/kuma/emision/EmisionFragment.kt
package knf.kuma.emision

import android.os.Bundle
import android.preference.PreferenceManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import knf.kuma.R
import knf.kuma.database.CacheDB
import knf.kuma.pojos.AnimeObject
import knf.kuma.widgets.emision.WEmisionProvider
import kotlinx.android.synthetic.main.recycler_emision.*
import org.jetbrains.anko.doAsync
import org.jsoup.Jsoup
import pl.droidsonroids.jspoon.Jspoon
import java.util.*

class EmisionFragment : Fragment() {
    private val dao = CacheDB.INSTANCE.animeDAO()
    private var adapter: EmisionAdapter? = null
    private var isFirst = true

    private val blacklist: Set<String>?
        get() = if (PreferenceManager.getDefaultSharedPreferences(context).getBoolean("show_hidden", false))
            LinkedHashSet()
        else
            PreferenceManager.getDefaultSharedPreferences(context).getStringSet("emision_blacklist", LinkedHashSet())

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return LayoutInflater.from(context).inflate(R.layout.recycler_emision, container, false)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        adapter = EmisionAdapter(this)
        recycler.adapter = adapter
        if (context != null)
            CacheDB.INSTANCE.animeDAO().getByDay(arguments!!.getInt("day", 1), blacklist!!).observe(this, Observer { animeObjects ->
                progress.visibility = View.GONE
                if (isFirst && animeObjects != null && animeObjects.isNotEmpty()) {
                    isFirst = false
                    adapter?.update(animeObjects as MutableList<AnimeObject>) { smoothScroll() }
                    recycler.scheduleLayoutAnimation()
                    checkStates(animeObjects)
                }
                error.visibility = if (animeObjects == null || animeObjects.isEmpty()) View.VISIBLE else View.GONE
            })
    }

    private fun smoothScroll() {
        //recycler.layoutManager?.smoothScrollToPosition(recycler,null,0)
    }

    private fun checkStates(animeObjects: MutableList<AnimeObject>?) {
        doAsync {
            try {
                for (animeObject in animeObjects!!) {
                    try {
                        val document = Jsoup.connect(animeObject.link).cookie("device", "computer").get()
                        val animeObject1 = AnimeObject(animeObject.link!!, Jspoon.create().adapter(AnimeObject.WebInfo::class.java).fromHtml(document.outerHtml()))
                        if (animeObject1.state != "En emisión") {
                            dao.updateAnime(animeObject1)
                            adapter?.remove(adapter!!.list.indexOf(animeObject))
                            WEmisionProvider.update(context!!)
                        }
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }

                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    internal fun reloadList() {
        if (context != null)
            CacheDB.INSTANCE.animeDAO().getByDay(arguments!!.getInt("day", 1), blacklist!!).observe(this, Observer { animeObjects ->
                error!!.visibility = View.GONE
                if (animeObjects != null && animeObjects.isNotEmpty())
                    adapter?.update(animeObjects as MutableList<AnimeObject>) { smoothScroll() }
                else
                    adapter?.update(ArrayList()) { smoothScroll() }
                if (animeObjects == null || animeObjects.isEmpty())
                    error!!.visibility = View.VISIBLE
            })
    }

    companion object {

        operator fun get(day: AnimeObject.Day): EmisionFragment {
            val bundle = Bundle()
            bundle.putInt("day", day.value)
            val fragment = EmisionFragment()
            fragment.arguments = bundle
            return fragment
        }
    }
}
